app/models/import_service_config.rb

```diff
@@ -0,0 +1,220 @@
+  end
+
+  def user
+    User.find_by(id: user_id)
+  end
+
+  def branding_group
+    BrandingGroup.find_by(id: branding_group_id)
+  end
+
+  # order of associations to walk for sourcing attributes
+  # should be overwritten in included class
+  def association_order
+    []
+  end
+
+  # walk association_order to populate associated IDs from a starting point
+  def traverse_associations!
+    association_order.each do |config|
+      entity_id = send(config[:id])
+      entity = ImportService.call_api_client(client, config[:method], entity_id)
+      associated_id = id_from(entity, config[:association])
+      setter = "#{config[:assoc_id]}=" # e.g. self.study_id = 'foo'
+      send(setter, associated_id)
+    end
+    attributes
+  end
+
+  # extract an ID from an association
+  def id_from(entity, ...)
+    client.send(id_from_method, entity, ...)
+  end
+
+  def taxon_from(value, attribute: :common_name)
+    Taxon.find_by(attribute.to_sym => value)
+  end
+
+  def taxon_names
+    defined?(self.class::PREFERRED_TAXONS) ? self.class::PREFERRED_TAXONS : []
+  end
+
+  # empty hash, will be overwritten in included class
+  def study_mappings
+    {}
+  end
+
+  # empty hash, will be overwritten in included class
+  def study_file_mappings
+    {}
+  end
+
+  # empty method for getting file access info; overwrite in included class
+  def file_access_info(...); end
+
+  # create a new SCP model instance from remote data
+  def to_scp_model(model, defaults, mappings, remote_data)
+    raise ArgumentError, "#{model} not in #{ALLOWED_SCP_MODELS.join(', ')}" unless ALLOWED_SCP_MODELS.include?(model)
+
+    model_attributes = defaults.dup
+    mappings.each do |scp_name, config_name|
+      if config_name.is_a?(Hash)
+        config_name.each do |embedded_doc_attr, nested_config_name|
+          model_attributes[scp_name] ||= {}
+          model_attributes[scp_name][embedded_doc_attr] = remote_data[nested_config_name]
+        end
+      else
+        model_attributes[scp_name] = remote_data[config_name]
+      end
+    end
+    model.new(**model_attributes)
+  end
+
+  def to_study
+    study_info = load_study.with_indifferent_access
+    study = to_scp_model(Study, study_default_settings, study_mappings, study_info)
+    study.name = sanitize_attribute(study.name)
+    study.description = sanitize_attribute(study.description)
+    study.external_identifier = study_id
+    study
+  end
+
+  def to_study_file(study_id, taxon_common_name, taxon_attribute: :common_name, format_attribute: :file_format)
+    file_info = load_file.with_indifferent_access
+    study_file = to_scp_model(StudyFile, study_file_default_settings, study_file_mappings, file_info)
+    # assign study, taxon, and content_type
+    study_file.study_id = study_id
+    study_file.taxon_id = taxon_from(taxon_common_name, attribute: taxon_attribute)&.id
+    ext = file_info[format_attribute].gsub(/^\./, '') # trim leading period, if present
```

Comment:
Out of curiosity, do you recall encountering any compound extensions (e.g. ".tar.gz") in NeMO or HCA?

Reply:
I've seen `.h5ad.tar` in NeMO, but not `.tar.gz`.  HCA I haven't seen any yet with compound names, but I haven't looked super hard for any.